### ISSUE TITLE
Inquiry about code in finetune.py

### ISSUE THREAD

**Issue by wendylongman2014** [tloen/alpaca-lora#327]
I am writing to inquire about a section of code in your project's finetune.py file. Specifically, I am referring to lines 256 to 261:

```python
old_state_dict = model.state_dict
model.state_dict = (
    lambda self, *_, **__: get_peft_model_state_dict(
        self, old_state_dict()
    )
).__get__(model, type(model))
```

I have noticed that this code is necessary for proper loss optimization, but I do not fully understand why. Could you please explain what this code does and why it is necessary for proper model training?

Thank you for your time and for maintaining this project.

**Comment by wendylongman2014** [comment]
In other words, why is it causing training errors when loading all parameter weights when continuing training from a checkpoint without performing such an operation?

**Comment by ElleLeonne** [comment]
This is script that binds any calls requesting to get_state_dict of the model, to instead perform get_peft_model_state_dict

However, it was in the repository prior to the recent peft update (which changed how the set_peft_state_dict call functioned) and some people seem to be getting errors with and without it.

I don't personally understand how LoRA works well enough to parse the new changes, but that should at least answer your question about how the binding script works and what it does.

**Comment by wendylongman2014** [comment]
Thank you for your reply, but I am still curious why this needs to be done. This will prevent the base model's weights from being loaded when continuing training. My experiments show that this is correct, but why can't we load those weights? Can the author of this code answer this? @tloen

**Comment by lywinged** [comment]
It's for load_best_model() in transformers trainer, because PyTorch doesn't support to load a quantized model from checkpoint into int8, so we need PyTorch's state_dict.

**Comment by wendylongman2014** [comment]
Thank you so much! You have solved my doubts and confusion.

**Comment by ndvbd** [comment]
Can you explain why you use:
`model.config.use_cache = False`

**Comment by lywinged** [comment]
> Can you explain why you use:
> `model.config.use_cache = False`

For saving memory

**Comment by ndvbd** [comment]
How does it save memory, and how much approx?

**Comment by lywinged** [comment]
> Thank you @lywinged, how does it save memory, and how much approx?

Because it will save the previous self-attn's hidden states for next token in decoder, when the sequence length is too long or the batch size is large, it cost memory. The memory usage depends on your input tokens number, model architecture and hyper parameter setting. So only you can tell how much. It's better to read the huggingface transformers doc if you want to use it.
